library/serialize.c

```diff
@@ -0,0 +1,333 @@
+ *   0..0       0x45    Type identifier
+ *   1..3       func    Function identifier
+ * Function identifiers are conventionally arranged as follows:
+ *   bits 23-16   Library code (1=sockets)
+ *   bits 15-8    Function identifier inside the library
+ *   bits 7-4     Number of parameters
+ *   bits 3-0     Number of results after status code on success
+ *
+ * Format of a result packet (by byte offset):
+ *   0..0       0x72    Type identifier
+ *   1..3       size    Number of bytes to follow
+ *   4..3+size  data    Content of the byte array
+ *
+ * All functions currently follow the following result status
+ * convention. The first result is a status code (MBEDTLS_ERR_xxx
+ * constant). If this is 0 then more results follow. If this is
+ * nonzero then it is the only result.
+ */
+
+#if !defined(MBEDTLS_CONFIG_FILE)
+#include "mbedtls/config.h"
+#else
+#include MBEDTLS_CONFIG_FILE
+#endif
+
+#include <limits.h>
+#include <stdint.h>
+#include <string.h>
+#include <signal.h>
+#include "mbedtls/serialize.h"
+
+#if defined(MBEDTLS_SERIALIZE_C)
+
+static int serialize_write_fd = -1;
+static int serialize_read_fd = -1;
+
+#if defined(MBEDTLS_SERIALIZE_FORK_FRONTEND_C)
+#include <unistd.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <signal.h>
+
+static int host_pid;
+
+static void kill_host_frontend( void )
+{
+    kill( host_pid, SIGHUP );
+}
+
+static void relay_signal( int signum )
+{
+    /* Let the atexit function kill the frontend */
+    exit( signum + 128 );
+}
+
+#define CHECK( expr, var )                      \
+    do {                                        \
+        var = ( expr );                         \
+        if( var < 0 )                           \
+        {                                       \
+            perror( #expr );                    \
+            goto cleanup;                       \
+        }                                       \
+    } while( 0 );
+static int mbedtls_serialize_prepare( void )
+{
+    int ret;
+    int null = -1;
+    int host_to_target[2] = {-1, -1}, target_to_host[2] = {-1, -1};
+    const char *const frontend_path[] = {
+        "programs/host/frontend",
+        "../programs/host/frontend",
+        "../../programs/host/frontend",
+    };
+    const char *frontend_exe = NULL;
+    size_t i;
+
+    /* Look for frontend program */
+    for( i = 0; i < sizeof( frontend_path ) / sizeof( *frontend_path ); i++ )
+    {
+        if( access( frontend_path[i], F_OK ) >= 0 )
+        {
+            frontend_exe = frontend_path[i];
+            break;
+        }
+    }
+    if( frontend_exe == NULL )
+    {
+        fprintf( stderr, "Host frontend executable for offloading not found!\n" );
+        return( -1 );
+    }
+
+    /* Prepare the plumbing */
+    CHECK( open( "/dev/null", O_RDONLY ), null );
+    CHECK( dup2( null, 3 ), ret );
+    CHECK( dup2( null, 4 ), ret );
+    if( null != 3 && null != 4 )
```

Comment:
Why are we hardcoding the file descriptors here? It 3 and 4 are already in use, they get closed by these calls. What about using `dup` which uses the lowest-numbered free descriptor?

Reply:
The reason for hard-coding descriptors 3 & 4 is that there is no interface between the forked child and the application (```frontend```) that will use the descriptors. In absence of forking ```frontend.c``` will need interface specific implementation to negotiate the descriptors. For this case i.e. forking the issue can be solved by providing an interface between child and app. 
I have add a new commit to introduce command line arguments to supply original pipe descriptors and remove use of dup. https://github.com/ARMmbed/mbedtls/pull/1029/commits/98ed19461a019e3d7e2ab7622876d13d5d34f154